packages/table/src/Row.tsx

```diff
@@ -0,0 +1,312 @@
+import { Transition } from 'react-transition-group';
+import IconButton from '@leafygreen-ui/icon-button';
+import Icon from '@leafygreen-ui/icon';
+import { isComponentType } from '@leafygreen-ui/lib';
+import { css, cx } from '@leafygreen-ui/emotion';
+import { uiColors } from '@leafygreen-ui/palette';
+import CheckboxCell from './CheckboxCell';
+import { useTableContext, Types, DataType } from './TableContext';
+
+const rowStyle = css`
+  cursor: pointer;
+  border-bottom: 1px solid ${uiColors.gray.light2};
+  color: ${uiColors.gray.dark2};
+
+  & > td > div {
+    height: 40px;
+    overflow: hidden;
+    transition: all 150ms ease-in-out;
+  }
+`;
+
+const altColor = css`
+  &:nth-of-type(even) {
+    background-color: ${uiColors.gray.light3};
+  }
+`;
+
+const iconButtonMargin = css`
+  margin-right: 4px;
+`;
+
+const disabledStyle = css`
+  background-color: ${uiColors.gray.light2};
+  color: ${uiColors.gray.base};
+  cursor: not-allowed;
+  border-bottom: 1px solid ${uiColors.gray.light1};
+`;
+
+const disabledCell = css`
+  border-top: 1px solid ${uiColors.gray.light1};
+`;
+
+const displayFlex = css`
+  display: flex;
+  align-items: center;
+`;
+
+const truncation = css`
+  max-width: 100px;
+  white-space: nowrap;
+  text-overflow: ellipsis;
+  overflow: hidden;
+`;
+
+const transitionStyles = {
+  default: css`
+    transition: border 150ms ease-in-out;
+    border-bottom-color: transparent;
+
+    & > td {
+      padding-top: 0px;
+      padding-bottom: 0px;
+    }
+
+    & > td > div {
+      max-height: 0;
+    }
+  `,
+
+  entered: css`
+    border-bottom-color: ${uiColors.gray.light2};
+
+    & > td > div {
+      max-height: 40px;
+    }
+  `,
+};
+
+const styleMap = {
+  left: [DataType.String, DataType.Weight, DataType.ZipCode, DataType.Date],
+  right: [DataType.Number, DataType.Quantity],
+};
+
+function styleColumn(index: string, dataType: DataType) {
+  let justify = '';
+
+  if (styleMap.left.includes(dataType)) {
+    justify = 'flex-start';
+  } else if (styleMap.right.includes(dataType)) {
+    justify = 'flex-end';
+  }
+
+  return css`
+    & td:nth-child(${index}) > div {
+      justify-content: ${justify};
+    }
+  `;
+}
+
+function getIndentLevelStyle(indentLevel: number) {
+  return css`
+    & > td:nth-child(2) {
+      padding-left: ${8 + indentLevel * 16}px;
+    }
+  `;
+}
+
+interface RowProps extends React.ComponentPropsWithoutRef<'tr'> {
+  expanded?: boolean;
+  disabled?: boolean;
+  indentLevel?: number;
+  isParentExpanded?: boolean;
+}
+
+const Row = React.forwardRef(
+  (
+    {
+      expanded = false,
+      disabled = false,
+      children,
+      className,
+      indentLevel = 0,
+      isParentExpanded = true,
+      ...rest
+    }: RowProps,
+    ref: React.Ref<any>,
+  ) => {
+    const {
+      state: { data, columnInfo, hasNestedRows, hasRowSpan, selectable },
+      dispatch,
+    } = useTableContext();
+
+    const indexRef = React.useRef(Math.floor(Math.random() * 1000000 + 1));
```

Comment:
So React will only use the value given from here the first time, but JavaScript will run this code everytime. This can be slow.

Let's instead move the logic into a function (defined outside the component) and I believe you can pass in a function to useRef to get the initial value: `useRef(generateNumber)`.

Reply:
👍